pkg/upgrade/config/v1alpha1/config.go

```diff
@@ -0,0 +1,199 @@
+/*
+Copyright 2019 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/upgrade/v1alpha1"
+)
+
+// UpgradeConfig is wrapper over apis.UpgradeConfig which is
+// upgrade config for a particular job.
+type UpgradeConfig struct {
+	object *apis.UpgradeConfig
+}
+
+// Builder helps to build UpgradeConfig instance
+type Builder struct {
+	*UpgradeConfig
+	checks map[*Predicate]string
+	errors []error
+}
+
+// Predicate abstracts conditional logic w.r.t the upgradeConfig instance
+//
+// NOTE:
+// Predicate is a functional approach versus traditional approach to mix
+// conditions such as *if-else* within blocks of business logic
+//
+// NOTE:
+// Predicate approach enables clear separation of conditionals from
+// imperatives i.e. actions that form the business logic
+type Predicate func(*UpgradeConfig) bool
+
+// AddCheckf adds the predicate as a condition to be validated against the
+// upgrade config instance and format the message string according to format specifier.
+// If only predicate and message string is provided, it will treat it as the
+// value for the corresponding predicate.
+func (b *Builder) AddCheckf(p Predicate, predicateMsg string, args ...interface{}) *Builder {
+	b.checks[&p] = fmt.Sprintf(predicateMsg, args...)
+	return b
+}
+
+// AddCheck adds the predicate as a condition to be validated against the
+// upgrade config instance
+func (b *Builder) AddCheck(p Predicate) *Builder {
+	return b.AddCheckf(p, "")
+}
+
+// AddChecks adds the provided predicates as conditions to be validated against
+// the upgrade config instance
+func (b *Builder) AddChecks(predicates ...Predicate) *Builder {
+	for _, check := range predicates {
+		b.AddCheck(check)
+	}
+	return b
+}
+
+// Build returns the final instance of UpgradeConfig
+func (b *Builder) Build() (*apis.UpgradeConfig, error) {
+	if len(b.errors) != 0 {
+		return nil, errors.Errorf("%v", b.errors)
+	}
+	err := b.validate()
+	if err != nil {
+		return nil, err
+	}
+	return b.UpgradeConfig.object, nil
+}
+
+// validate will run checks against UpgradeConfig instance
+func (b *Builder) validate() error {
+	for cond := range b.checks {
+		pass := (*cond)(b.UpgradeConfig)
+		if !pass {
+			b.errors = append(b.errors,
+				errors.Errorf("%v", b.checks[cond]))
```

Comment:
can we rewrite this as:
```go
b.errors = append(b.errors, errors.Errorf("failed to validate: %s", b.checks[cond])
```

Reply:
Let validate return validate specific errors.
It should also append it to b.errors.
However there can be a local []error that validate can keep on appending.
The return of this validate should be this local []error
Let me know if I am clear?

Reply:
I want us to isolate validate specific errors from builder errors.
Builder errors can have all including validate errors